File: frontend/src/lib/polegadas.ts
export function normalizePolegadaInput(input: string): string {
  const raw = (input || '').trim().replace(/"/g, '');
  if (!raw) return '';
  const normalized = raw.replace(/\s+/g, ' ').replace('-', '.').replace(' ', '.');
  return normalized.endsWith('"') ? normalized : `${normalized}"`;
}

export function parsePolegadaToDecimal(input: string): number | null {
  const txt = (input || '').trim().replace(/"/g, '').replace(',', '.').replace('-', '.').replace(' ', '.');
  if (!txt) return null;
  const parts = txt.split('.').filter(Boolean);
  if (parts.length === 1) {
    if (parts[0].includes('/')) {
      const [n, d] = parts[0].split('/');
      const num = Number(n);
      const den = Number(d);
      if (!Number.isFinite(num) || !Number.isFinite(den) || den === 0) return null;
      return num / den;
    }
    const n = Number(parts[0]);
    return Number.isFinite(n) ? n : null;
  }
  if (parts.length === 2 && parts[1].includes('/')) {
    const i = Number(parts[0]);
    const [n, d] = parts[1].split('/');
    const num = Number(n);
    const den = Number(d);
    if (!Number.isFinite(i) || !Number.isFinite(num) || !Number.isFinite(den) || den === 0) return null;
    return i + num / den;
  }
  const n = Number(txt);
  return Number.isFinite(n) ? n : null;
}

export function decimalToMm(decimal: number): number {
  return decimal * 25.4;
}

export function formatMm(value: number): string {
  return value.toFixed(2).replace('.', ',');
}

export function buildPolegadaAliases(input: string, decimal: number): string[] {
  const mm = decimalToMm(decimal);
  const normalized = normalizePolegadaInput(input);
  const base = normalized.replace(/"/g, '');
  const aliases = new Set<string>([
    normalized,
    base,
    base.replace('.', ' '),
    base.replace('.', '-'),
    decimal.toString(),
    decimal.toString().replace('.', ','),
    mm.toFixed(2),
    mm.toFixed(2).replace('.', ','),
  ]);
  if (base === '1/2') {
    aliases.add('meia');
    aliases.add('meio');
    aliases.add('0.5');
    aliases.add('0,5');
  }
  return [...aliases].filter(Boolean);
}

File: backend/apps/produtos/polegadas.py
from __future__ import annotations

import re
from decimal import Decimal, InvalidOperation


def parse_polegada_to_decimal(raw: str) -> Decimal | None:
    txt = (raw or '').strip().replace('"', '').replace(',', '.')
    if not txt:
        return None
    txt = txt.replace('-', '.').replace(' ', '.')
    parts = [p for p in txt.split('.') if p]
    try:
        if len(parts) == 1:
            if '/' in parts[0]:
                n, d = parts[0].split('/', 1)
                return Decimal(n) / Decimal(d)
            return Decimal(parts[0])
        if len(parts) == 2 and '/' in parts[1]:
            i = Decimal(parts[0])
            n, d = parts[1].split('/', 1)
            return i + (Decimal(n) / Decimal(d))
        return Decimal(txt)
    except (InvalidOperation, ZeroDivisionError, ValueError):
        return None


def normalize_polegada_label(raw: str) -> str:
    dec = parse_polegada_to_decimal(raw)
    if dec is None:
        txt = (raw or '').strip()
        return txt if txt.endswith('"') else f'{txt}"' if txt else ''
    base = (raw or '').strip().replace('"', '')
    if '/' in base:
        base = base.replace('-', '.').replace(' ', '.')
        parts = [p for p in base.split('.') if p]
        if len(parts) == 2 and '/' in parts[1]:
            return f'{parts[0]}.{parts[1]}"'
        if len(parts) == 1:
            return f'{parts[0]}"'
    norm = f'{dec.normalize()}'
    return f'{norm}"'


def aliases_for_polegada(descricao: str, valor_decimal: Decimal, valor_mm: Decimal) -> list[str]:
    aliases = {
        (descricao or '').strip(),
        (descricao or '').replace('"', '').strip(),
        str(valor_decimal),
        str(valor_decimal).replace('.', ','),
        f'{valor_mm:.2f}',
        f'{valor_mm:.2f}'.replace('.', ','),
    }
    desc_clean = (descricao or '').strip().lower()
    if desc_clean in {'1/2"', '1/2'}:
        aliases.update({'meia', 'meio', '0.5', '0,5'})
    aliases = {a for a in aliases if a}
    return sorted(aliases)


def extract_mm_from_term(term: str) -> Decimal | None:
    txt = (term or '').strip().replace(',', '.')
    if not txt:
        return None
    if re.fullmatch(r'\d+(\.\d+)?', txt):
        try:
            return Decimal(txt)
        except InvalidOperation:
            return None
    return None
